eth/backend.go

```diff
@@ -158,29 +160,116 @@ func New(stack *node.Node, config *Config) (*Ethereum, error) {
+		if config.ExternalSnapshotDownloaderAddr != "" {
+			cli, cl, innerErr := snapshotsync.NewClient(config.ExternalSnapshotDownloaderAddr)
+			if innerErr != nil {
+				return nil, innerErr
+			}
+			defer cl() //nolint
+
+			_, innerErr = cli.Download(context.Background(), &snapshotsync.DownloadSnapshotRequest{
+				NetworkId: config.NetworkID,
+				Type:      config.SnapshotMode.ToSnapshotTypes(),
+			})
+			if innerErr != nil {
+				return nil, innerErr
+			}
 
-		snapshotKV := chainDb.KV()
-		snapshotKV, err = torrent.WrapBySnapshots(snapshotKV, dbPath, config.SnapshotMode)
-		if err != nil {
-			return nil, err
-		}
-		chainDb.SetKV(snapshotKV)
-		err = torrent.PostProcessing(chainDb, config.SnapshotMode)
-		if err != nil {
-			return nil, err
+			waitDownload := func() (map[snapshotsync.SnapshotType]*snapshotsync.SnapshotsInfo, error) {
+				snapshotReadinessCheck := func(mp map[snapshotsync.SnapshotType]*snapshotsync.SnapshotsInfo, tp snapshotsync.SnapshotType) bool {
+					if mp[tp].Readiness != int32(100) {
+						log.Info("Downloading", "snapshot", tp, "%", mp[tp].Readiness)
+						return false
+					}
+					return true
+				}
+				for {
+					mp := make(map[snapshotsync.SnapshotType]*snapshotsync.SnapshotsInfo)
+					snapshots, err1 := cli.Snapshots(context.Background(), &snapshotsync.SnapshotsRequest{NetworkId: config.NetworkID})
+					if err1 != nil {
+						return nil, err1
+					}
+					for i := range snapshots.Info {
+						if mp[snapshots.Info[i].Type].SnapshotBlock < snapshots.Info[i].SnapshotBlock && snapshots.Info[i] != nil {
+							mp[snapshots.Info[i].Type] = snapshots.Info[i]
+						}
+					}
+
+					downloaded := true
+					if config.SnapshotMode.Headers {
+						if !snapshotReadinessCheck(mp, snapshotsync.SnapshotType_headers) {
+							downloaded = false
+						}
+					}
+					if config.SnapshotMode.Bodies {
+						if !snapshotReadinessCheck(mp, snapshotsync.SnapshotType_bodies) {
+							downloaded = false
+						}
+					}
+					if config.SnapshotMode.State {
+						if !snapshotReadinessCheck(mp, snapshotsync.SnapshotType_state) {
+							downloaded = false
+						}
+					}
+					if config.SnapshotMode.Receipts {
+						if !snapshotReadinessCheck(mp, snapshotsync.SnapshotType_receipts) {
+							downloaded = false
+						}
+					}
+					if downloaded {
+						return mp, nil
+					}
+					time.Sleep(time.Second * 10)
+				}
+			}
+			downloadedSnapshots, innerErr := waitDownload()
+			if innerErr != nil {
+				return nil, innerErr
+			}
+			snapshotKV := chainDb.KV()
+
+			snapshotKV, innerErr = snapshotsync.WrapBySnapshots2(snapshotKV, downloadedSnapshots)
+			if innerErr != nil {
+				return nil, innerErr
+			}
+			chainDb.SetKV(snapshotKV)
+			innerErr = snapshotsync.PostProcessing(chainDb, config.SnapshotMode)
+			if innerErr != nil {
+				return nil, innerErr
+			}
+		} else {
```

Comment:
Does this mean you are keeping the functionality of in-process downloading still?

Reply:
Or is this seeding?